include/ocpp/common/message_queue.hpp

```diff
@@ -162,6 +191,17 @@ template <typename M> class MessageQueue {
         return false;
     }
 
+    bool allowed_to_send_message(const std::shared_ptr<ControlMessage<M>> message) {
+        if (message->stall_until_accepted and !this->is_registration_status_accepted) {
+            return false;
+        }
+
+        if (message->timestamp > DateTime()) {
+            return false;
+        }
+        return true;
+    }
```

Comment:
Again, I would write this as a free function and pass `is_registration_status_accepted` as a parameter.  Furthermore, `ControlMessage<M>` shouldn't be passed as a shared pointer, but by reference.
If written as a free function, you would need to write it as template (because of `ControlMessage<M>`).
Performance wise, `DateTime()` might not be instantiated and evaluated each and every time.